### PR TITLE
fix(sequelize/query): release connection after retry

### PR DESCRIPTION
**Pull request by lext-7**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in CONTRIBUTING.md?

### Description of change

Connections should be release after retry rather than being released after every query.
It will cause `'Resource not currently part of this pool'` error if it's released after every query.

In v3, sequelize use `coopernurse/node-pool@2.4.8` which won't raise a error if a connection is released multiple times. See here [https://github.com/coopernurse/node-pool/blob/v2.4/lib/generic-pool.js#L420](https://github.com/coopernurse/node-pool/blob/v2.4/lib/generic-pool.js#L420)

But in v4, sequelize use new version of `coopernurse/node-pool` which will return a rejected promise if a connection is released multiple times. See here [https://github.com/coopernurse/node-pool/blob/master/lib/Pool.js#L422](https://github.com/coopernurse/node-pool/blob/master/lib/Pool.js#L422).

Because of this error, if you are using retry, you will get `'Resource not currently part of this pool'` when got every sql error. You can not get a normal sql error which shows you what is wrong.

### CODE REVIEW

File: lib/sequelize.js [hunk 1]
```diff
@@ -529,11 +529,12 @@ class Sequelize {
     }).then(connection => {
       const query = new this.dialect.Query(connection, this, options);
 
-      return retry(() => query.run(sql, bindParameters).finally(() => {
-        if (!options.transaction) {
-          return this.connectionManager.releaseConnection(connection);
-        }
-      }), Utils._.extend(this.options.retry, options.retry || {}));
+      return retry(() => query.run(sql, bindParameters), Utils._.extend(this.options.retry, options.retry || {}))
+        .finally(() => {
```
Comment (sushantdhiman) [inline lib/sequelize.js]: You can move this to `finally` block below

Comment (lext-7) [reply lib/sequelize.js]: But `connection` is not defined in that `finally` block below.

Comment (sushantdhiman) [reply lib/sequelize.js]: Right :), then just add a unit test and it should be ready to go

File: test/integration/sequelize.test.js [hunk 2]
```diff
@@ -728,6 +728,23 @@ describe(Support.getTestDialectTeaser('Sequelize'), () => {
...
+          // retry for all errors
+          match: null
+        }
+      }).catch(err => {
+        // If the connection is released multiple times,
+        // than an Error(not Sequelize BaseError) will be raised by generic-pool.
+        // If the connection is released only once,
+        // then a DatabaseError which inherits Sequelize BaseError will be raised,
+        // because this is a wrong sql.
+        expect(err).to.be.an.instanceof(this.sequelize.Error);
```
Comment (sushantdhiman) [inline test/integration/sequelize.test.js]: Add a spy on `this.sequelize.connectionManager.releaseConnection`, assert against number of calls

File: lib/sequelize.js [hunk 3]
```diff
@@ -529,11 +529,12 @@ class Sequelize {
     }).then(connection => {
       const query = new this.dialect.Query(connection, this, options);
 
-      return retry(() => query.run(sql, bindParameters).finally(() => {
-        if (!options.transaction) {
-          return this.connectionManager.releaseConnection(connection);
-        }
-      }), Utils._.extend(this.options.retry, options.retry || {}));
+      return retry(() => query.run(sql, bindParameters), Utils._.extend(this.options.retry, options.retry || {}))
```
Comment (sushantdhiman) [inline lib/sequelize.js]: not easy to read, add a new `const retryOptions` along side `const query`

File: test/integration/sequelize.test.js [hunk 4]
```diff
@@ -728,6 +728,23 @@ describe(Support.getTestDialectTeaser('Sequelize'), () => {
       });
     });
 
+    it('connections should be released only once', function() {
```
Comment (sushantdhiman) [inline test/integration/sequelize.test.js]: `connection should be released only once when retry fails`

**Pull request merged**